crates/ironrdp-connector/src/credssp.rs

```diff
@@ -96,18 +99,42 @@ impl CredsspSequence {
         kerberos_config: Option<KerberosConfig>,
     ) -> ConnectorResult<(Self, credssp::TsRequest)> {
         let config = &connector.config;
-        if let crate::Credentials::SmartCard { .. } = config.credentials {
-            return Err(general_err!(
-                "CredSSP with smart card credentials is not currently supported"
-            ));
-        }
-
-        let username = Username::new(config.credentials.username(), config.domain.as_deref())
-            .map_err(|e| custom_err!("invalid username", e))?;
-
-        let credentials = sspi::AuthIdentity {
-            username,
-            password: config.credentials.secret().to_owned().into(),
+        let credentials: sspi::Credentials = match &config.credentials {
+            Credentials::UsernamePassword { username, password } => {
+                let username = Username::new(username, config.domain.as_deref())
+                    .map_err(|e| custom_err!("invalid username", e))?;
+
+                sspi::AuthIdentity {
+                    username,
+                    password: password.to_owned().into(),
+                }
+                .into()
+            }
+            Credentials::SmartCard { pin, config } => match config {
+                Some(config) => {
+                    let cert: Certificate = picky_asn1_der::from_bytes(&config.certificate)
+                        .map_err(|_e| general_err!("can't parse certificate"))?;
+                    let key = PrivateKey::from_pkcs1(&config.private_key)
+                        .map_err(|_e| general_err!("can't parse private key"))?;
+                    let identity = sspi::SmartCardIdentity {
+                        username: extract_user_principal_name(&cert)
+                            .or_else(|| extract_user_name(&cert))
+                            .unwrap_or_default(),
+                        certificate: cert,
+                        reader_name: config.reader_name.clone(),
+                        card_name: None,
+                        container_name: config.container_name.clone(),
+                        csp_name: config.csp_name.clone(),
+                        pin: pin.as_bytes().to_vec().into(),
+                        private_key_file_index: None,
+                        private_key: Some(key.into()),
+                    };
+                    sspi::Credentials::SmartCard(Box::new(identity))
```

Comment:
thought: I think we should not expose picky types in sspi-rs public API. I may change that before releasing sspi v0.12.0 :thinking: